buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java

```diff
@@ -0,0 +1,273 @@
+            dependencies.add(distribution.configuration.getName(), dependencyNotation(project, distribution));
+
+            // no extraction allowed for rpm or deb
+            if (distribution.getType() != Type.RPM && distribution.getType() != Type.DEB) {
+                // for the distribution extracted, add a root level task that does the extraction, and depend on that
+                // extracted configuration as an artifact consisting of the extracted distribution directory
+                dependencies.add(distribution.getExtracted().configuration.getName(),
+                    projectDependency(project, ":", configName("extracted_elasticsearch", distribution)));
+                // ensure a root level download task exists
+                setupRootDownload(project.getRootProject(), distribution);
+            }
+        }
+    }
+
+    private void setupRootDownload(Project rootProject, ElasticsearchDistribution distribution) {
+        String extractTaskName = extractTaskName(distribution);
+        // NOTE: this is *horrendous*, but seems to be the only way to check for the existence of a registered task
+        try {
+            rootProject.getTasks().named(extractTaskName);
+            // already setup this version
+            return;
+        } catch (UnknownTaskException e) {
+            // fall through: register the task
+        }
+        setupDownloadServiceRepo(rootProject);
+
+        final ConfigurationContainer configurations = rootProject.getConfigurations();
+        String downloadConfigName = configName("elasticsearch", distribution);
+        String extractedConfigName = "extracted_" + downloadConfigName;
+        final Configuration downloadConfig = configurations.create(downloadConfigName);
+        configurations.create(extractedConfigName);
+        Object distroDep = dependencyNotation(rootProject, distribution);
+        rootProject.getDependencies().add(downloadConfigName, distroDep);
+
+        // add task for extraction, delaying resolving config until runtime
+        if (distribution.getType() == Type.ARCHIVE || distribution.getType() == Type.INTEG_TEST_ZIP) {
+            Supplier<File> archiveGetter = downloadConfig::getSingleFile;
+            String extractDir = rootProject.getBuildDir().toPath().resolve("elasticsearch-distros").resolve(extractedConfigName).toString();
+            TaskProvider<Copy> extractTask = rootProject.getTasks().register(extractTaskName, Copy.class, copyTask -> {
+                copyTask.dependsOn(downloadConfig);
+                copyTask.doFirst(t -> rootProject.delete(extractDir));
+                copyTask.into(extractDir);
+                copyTask.from((Callable<FileTree>)() -> {
+                    File archiveFile = archiveGetter.get();
+                    String archivePath = archiveFile.toString();
+                    if (archivePath.endsWith(".zip")) {
+                        return rootProject.zipTree(archiveFile);
```

Comment:
maybe it's just me, but I think `rootProject.zipTree(downloadConfig.getSingleFile())` would be much easier to read.  this might be subjective, but I find that variables only used once or a few times but really far apart make the code significantly harder to read.

Reply:
I had something like that before, but Mark suggested switching on the extension. I don't have a strong preference, other than this is already in and works.

Reply:
We were switching based on hard-coded distribution types. That certainly worked but is more brittle than just looking at the actual file-extension. That also was using raw strings before the conversion to enums which was even more brittle. Other option is to have some property on the distribution to indicate what _kind_ of archive it is and key off of that.